gno.land/pkg/sdk/vm/keeper.go

```diff
@@ -136,6 +145,92 @@ func (vm *VMKeeper) getGnoStore(ctx sdk.Context) gno.Store {
 	}
 }
 
+// Namespace can be either a user or crypto address.
+// XXX: Uppercase should be valid but transform into lowercase ?
+var reNamespace = regexp.MustCompile(`^gno.land/(?:r|p)/([a-zA-Z]+[_a-zA-Z0-9]+)`)
+
+// checkNamespacePermission check if the user as given has correct permssion to on the given pkg path
+func (vm *VMKeeper) checkNamespacePermission(ctx sdk.Context, creator crypto.Address, pkgPath string) error {
+	const sysUsersPkg = "gno.land/r/demo/users"
+
+	store := vm.getGnoStore(ctx)
+
+	// if `sysUsersPkg` does not exists -> skip validation.
+	usersPkg := store.GetPackage(sysUsersPkg, false)
+	if usersPkg == nil {
+		return nil
+	}
+
+	// XXX: is this necessary ?
+	pkgPath = filepath.Clean(pkgPath) // cleanup pkgpath
+
+	match := reNamespace.FindStringSubmatch(pkgPath)
+	if len(match) != 2 {
+		return fmt.Errorf("%w: %q", ErrInvalidRealm, pkgPath)
+	}
+	username := match[1]
+
+	// Lowercase username
+	username = strings.ToLower(username)
+
+	// Allow user with their own address as namespace
+	if addr, err := crypto.AddressFromBech32(username); err == nil {
+		if addr.Compare(creator) == 0 {
+			return nil
+		}
+
+		return fmt.Errorf("%w: %q", ErrUnauthorizedUser, username)
+	}
+
+	// Parse and run the files, construct *PV.
+	pkgAddr := gno.DerivePkgAddr(pkgPath)
+	msgCtx := stdlibs.ExecContext{
+		ChainID:       ctx.ChainID(),
+		Height:        ctx.BlockHeight(),
+		Timestamp:     ctx.BlockTime().Unix(),
+		OrigCaller:    creator.Bech32(),
+		OrigSendSpent: new(std.Coins),
+		OrigPkgAddr:   pkgAddr.Bech32(),
+		// XXX: should we remove the banker ?
+		Banker:      NewSDKBanker(vm, ctx),
```

Comment:
Since we're using `m.Eval`, which is read-only, we should not use the real banker here. The realm banker may not be aware of the read-only context and could inadvertently apply changes.